[FIX] account: if there is no invoice_tax_id and no refund_tax_id the…

wtaferner opened the pull request:
**Description of the issue/feature this PR addresses:**
If there is no `invoice_tax_id` and no `refund_tax_id` there must be a `company_id` at least

As on inverse set of repartition lines while unlinking the company check is executed, it is impossible to unset the link to `account.tax`
Based on this we assume that the best possible company would be the company of the current user...


**Current behavior before PR:**
Company Check is blocking and there is IMO no situation where the company should be false in account.tax in multi company

**Desired behavior after PR is merged:**
Company Check is passing

Info: @wt-io-it


--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/67701).

wtaferner commented:
@william-andre 
Uhh, nice, I like the automation of requesting a review. Now I do get a quick answer and collaboration 😉

MiquelRForgeFlow commented on addons/account/models/account_tax.py:
```diff
@@ -675,7 +675,7 @@ def _compute_factor(self):
     def _compute_company(self):
         for record in self:
-            record.company_id = record.invoice_tax_id and record.invoice_tax_id.company_id.id or record.refund_tax_id.company_id.id
+            record.company_id = record.invoice_tax_id and record.invoice_tax_id.company_id.id or record.refund_tax_id and record.refund_tax_id.company_id.id or self.env.user.company_id.id
```
```suggestion
            record.company_id = record.invoice_tax_id and record.invoice_tax_id.company_id.id or record.refund_tax_id and record.refund_tax_id.company_id.id or self.env.company.id
```

wtaferner replied:
Thx 👍🏼 I thought about that after pushing 😄 
I will fix that, in case there is no other objection by @william-andre or anyone at Odoo

william-andre reviewed (changes requested):
@wtaferner You need to rebase your branch for the failing test (it is a deterministic error that started a few days/weeks ago because of a hardcoded date in the setup of the test)

However, retrieving the company from the context is not correct and is likely to fail when doing inter-company flows.

Checking on the linked account would be better. But in which case do you have a repartition line not linked to a tax?

wtaferner commented:
@william-andre 
Ok, I will do.

About the use case, actually I try to reinitialize the chart template and during this process it does the magic itself, but basically it is based on some inverse field logic as far as I have identified to set the invoice_tax_id to False which triggers the issue.

Inter company flows in account...is there a tax or repartition line which is not exclusive to one company (basically no company set)?
The check itself is much appreciated as I suffered or avoided enough multi company because of these inconsistencies 😉

wtaferner commented:
@william-andre 
After testing without the user company context, I removed that magic and added an explicit `False` to the code and it should be now ready to be merged.

Nevertheless, I would be interested in an answer where a repartition line `company_id` would be `False` in an Odoo multi company context, if you know one explicitly instead of a generic thinking?

Thank you 🙏🏼

MiquelRForgeFlow commented on addons/account/models/account_tax.py:
```diff
@@ -675,7 +675,12 @@ def _compute_factor(self):
     @api.depends('invoice_tax_id.company_id', 'refund_tax_id.company_id')
```
```suggestion
    @api.depends('invoice_tax_id.company_id', 'refund_tax_id.company_id','account_id.company_id')
```

wtaferner replied:
Thx, this one makes a lot of sense.

william-andre commented on addons/account/models/account_tax.py:
```diff
@@ -675,7 +675,12 @@ def _compute_factor(self):
     def _compute_company(self):
         for record in self:
-            record.company_id = record.invoice_tax_id and record.invoice_tax_id.company_id.id or record.refund_tax_id.company_id.id
+            record.company_id = (
+                record.invoice_tax_id and record.invoice_tax_id.company_id.id or
+                record.refund_tax_id and record.refund_tax_id.company_id.id or
+                record.account_id and record.account_id.company_id.id or
+                False
+            )
```
If there is no `record.invoice_tax_id`, it will return an empty recordset. And calling the getter of a field on an epty recordset will return another empty recordset of the comodel or `False` for static values (`int`, `float`, `date`,...).

So you can do `record.invoice_tax_id.company_id` without needing to check  that `record.invoice_tax_id` is not falsy.

You also don't need to call `.id` to assign a recordset. It is working but I personally find it weird to assign an id and to retrieve a recordset.

```suggestion
            record.company_id = (
                record.invoice_tax_id.company_id
                or record.refund_tax_id.company_id
                or record.account_id.company_id
            )
```

I also don't understand why we are checking `invoice_tax_id` and `refund_tax_id` instead of `tax_id` directly. Maybe is it due to an ORM limitation?

This could be simplified (but arguably less readable) to
```python
            record.company_id = (record.tax_id or record.account_id).company_id
```

But really what bothers me is that it makes no sense to me to have a tax repartitions line that is not linked to any tax.
It is like having an invoice line not linked to an invoice... Can you explain your flow in more details? Because if this happens, it is probably because of a bug elsewhere.

william-andre replied:
By the way, the current code already does what you want.
```python
record.company_id = record.invoice_tax_id and record.invoice_tax_id.company_id.id or record.refund_tax_id.company_id.id
```
Is the same as
```python
if record.invoice_tax_id:
    record.company_id = record.invoice_tax_id.company_id.id or record.refund_tax_id.company_id.id
    # record.invoice_tax_id is not empty because of the if clause, 
    # record.invoice_tax_id.company_id is empty not empty because it is a required field
    # so company_id is set
else:
    record.company_id = record.refund_tax_id.company_id.id
    # if record.refund_tax_id is empty, then record.refund_tax_id.company_id,
    # then record.refund_tax_id.company_id.id := False
    # ==> same result as you
```

wtaferner replied:
Right, I agree, the check is not needed, I only followed what was done before to keep the initial PR diff small.
So, I would change that too...for the incident of how this is being triggered, I already said that it is during re-initializing the chart template again (which is not implemented in Odoo which is a pity).

During this the system tries to "unlink" the `invoice_tax_id` on a write whereas at the same time the `account_id` is still present. It was a huge pain to even find out that the source of the issue was in this method (which I do claim as wrong implementation not checking the full chain for a computed field `company_id`).

Maybe the `check_company` could be refined and improved, but having a computed `company_id` is probably the issue when you do block yourself. 

And well, I am pretty sure there are hundred bugs or more elsewhere, but I would prefer if we could fix one at a time 😉 and I hope we agree now that this is a bug and I happy we have found a common ground to get rid of it, right? 😄

wtaferner replied:
> By the way, the current code already does what you want.

No, it doesn't as the `account_id` is still set and has a company other than False set while the company check is running and it is only running for those records which are still linked.

william-andre replied:
> I already said that it is during re-initializing the chart template again (which is not implemented in Odoo which is a pity).

You can change the chart template if you don't have journal entries yet. If you have some, you should archive the taxes and accounts instead.

> During this the system tries to "unlink" the `invoice_tax_id` on a write whereas at the same time the `account_id` is still present.

It is already done in here https://github.com/odoo/odoo/blob/14.0/addons/account/models/chart_template.py#L205-L226 and it works without any issue; I'd like to know why it doesn't work in you case before merging. I just tried it in version 14.0 with `l10n_generic_coa`, `l10n_au`, `l10n_be` and `l10n_ch`, I could switch the template easily and no error was raised.

wtaferner replied:
I do not want to switch, I want to update the chart of account (accounts and taxes) based on an updated templates and the related xml ids algorithm, but nevertheless, this was just my entry point to end up here.

So, maybe take a step back look at the code and ask yourself if the implementation now does resolve what I have described.

The computed `company_id` must always be in line with ALL linked records (not just two out of three), otherwise the check company will blow up when passing an inconsistent state unrelated why this might happen, right?

william-andre replied:
But you should never be in the case where the repartition line is not linked to a tax. If you unlink the tax, the repartition line should be unlinked with it; hence no company check is done. If you change the tax (really bad idea), the company should not be changed and the check passes.

If you have to update the accounts and taxes, the correct way to do it is to archive/deprecate the records and create new ones. Or some fields can be changed in place, but then you don't need to unlink anything.

If a tax is not used, you can unlink it without merging this PR since it works when switching the CoA.

wtaferner replied:
Why do we discuss about the use case, when the consistency is not given on the one hand and enforced on the other (check company) and we could easily provide it by merging this PR?

One thing is sure, I would not have ended up doing this PR, if there was no issue at all.

wtaferner replied:
https://github.com/odoo/odoo/blob/14.0/addons/account/models/chart_template.py#L1066

This is IMO the culprit for the remove and add and somewhere the ORM (magic commands) is trying to unlink the `tax_invoice_id` before also removing the `account_id`.
Anyway, I would appreciate to also accept useful code just f

william-andre replied:
You should not fix the symptoms, but you should understand and fix the problem.

Having a repartition line not linked to a tax is a consistency problem that you/we should fix. Not hiding the problem like it is done here.

While this PR is not really wrong and could be merged, it certainly is not right either.

If you give me a way to reproduce, I can look into it if you want; but this really doesn't seem like the right way to fix it.

wtaferner replied:
Well, I do agree in general, but for me it is the source of the issue, but the reproduction is rather complex and you could not even follow my explanation and were pushing on some use cases which are for sure only one minimal PoV.

It is for sure best reproduced on an ORM level, but I have not more time to explain it better than I already did.
So, basically we are now stale and you do not agree that the `account_id` should be part of the equation in a computed `company_id` field against which it is compared on every write (with inverse and x2m command magic).

In the end you deny the possibility that it might happen at some point (unrelated whether or not I do now provide you an easy and clear reproduction from the use cases implemented in Odoo Core whereas I am just reusing some parts of it FYI).

Maybe you can talk to some ORM Framework specialist to analyse if it is just fixing the symptom or a real issue here...for me it is quite obvious that the computation is incomplete.

Anyway, I give up here and let you close it with a good feeling about it is the right thing to do.

Thank you for the discourse and your feedback ❤️
